### ISSUE TITLE
Find associated events for CSS Animations

### ISSUE THREAD

**Issue by benshayden** [catapult-project/catapult#1433]
From #1430 
I think this will require adding new flow events to blink, though I'm not sure exactly where.

**Comment by dstockwell** [comment]
As a start I think we  can add flow events in the same way that the async spans are generated.

**Comment by dj2** [comment]
@dstockwell do you know how would be best suited to add the needed events?

**Comment by dstockwell** [comment]
@dj2 I can add them, what exactly is necessary for the IR finder?

**Comment by dj2** [comment]
@benshayden can you let Doug know what you need from him?

**Comment by benshayden** [comment]
I need flow events from the Animation async slices to whatever trace events are directly caused by the animations. I don't know how CSS Animations work, so I don't know what slices would be caused directly by animations.
Basically, when you take a trace of an animation in chrome://tracing, then you should be able to see chains of spline-arrows from the Animation async slices in the Renderer process to the paint and frame events and whatever else is caused directly or indirectly by the animation.

I can make the finder follow the flow events and add the flowed-to slices to the Animation IR's associatedEvents.

For bonus points if you're up for it while you're in there: a flow event from the Document::updateStyle (or whatever else that can cause css animations) to the Animation async slice. The IR finder probably would not use this flow event, but it might be neat to be able to see when a Response causes a CSS Animation.

Thank you in advance!

**Comment by benshayden** [comment]
Sorry, I meant that the FlowEvents should have the same id as the Animation async slice, like how FlowEvents have the same id as latencyInfo async slices. Of course the flow events should start and end at ThreadSlices. I'm planning on using basically the same logic as InputLatencyAsyncSlice, insofar as it works for CSS Animations.
https://github.com/catapult-project/catapult/blob/master/tracing/tracing/extras/chrome/cc/input_latency_async_slice.html
 So, in tracing/extras/chrome, I'll register an AsyncSlice subclass named Animation to handle async slices titled "Animation" in the category "blink.animations", and override its associatedEvents getter with logic based on InputLatencyAsyncSlice.associatedEvents.

**Comment by benshayden** [comment]
Maybe like this: https://codereview.chromium.org/1338633004

**Comment by dstockwell** [comment]
Tried to add this, but there's no `INTERNAL_TRACE_EVENT_ADD_SCOPED_WITH_FLOW` in blink. It failed horribly when I tried to hack it in.

**Comment by dstockwell** [comment]
@dj2, @vmpstr, Any documentation on flow events?

Is it preferred to use  the *with_flow over flow_begin/step/end? I ask because there's no ...with_flow_begin/end yet.

What happens when I flow in our out multiple events with the same flow id?

**Comment by dj2** [comment]
flow_begin/step/end is deprecated. The with_flow variants are the new flow v2 and provide more flexibility.

The design doc is at: https://docs.google.com/document/d/1La_0PPfsTqHJihazYhff96thhjPtvq1KjAUOJu0dvEg/edit#heading=h.mlpdir2vdooz

**Comment by dstockwell** [comment]
@dj2 OK, I'll prefer with_flow. But in one case I was trying to convert there's an existing TRACE_EVENT_BEGIN/END pair and I can't see how to convert that.

**Comment by dj2** [comment]
Link?

**Comment by dstockwell** [comment]
eg. Document::updateLayoutTree: https://chromium.googlesource.com/chromium/src/+/master/third_party/WebKit/Source/core/dom/Document.cpp#1743

**Comment by dj2** [comment]
I'd be cautious modifying any of the devtools.timeline events without talking to them first as they use that for their timeline.

**Comment by benshayden** [comment]
@dstockwell Still working on this?

In Document::updateLayoutTree, could we add a new trace event like this?
TRACE_EVENT_WITH_FLOW0("blink", "Document::updateLayoutTree", animationId, FLOW_IN);

**Comment by dstockwell** [comment]
@benshayden, haven't made any progress here. This was waiting on an explanation of how to use flow events correctly. So we should discuss further.

**Comment by benshayden** [comment]
We might be able to use @skyostil 's Trace Contexts (frameblamr) to simplify this attribution problem as well as loading.
Until frameblamr is sorted out, though, maybe we could keep trying to get flow events working here.
Unfortunately, my knowledge of the corner cases of flow events is limited, and the design doc doesn't explain how to use them in this case.
For example, how does trace-viewer handle multiple flow events with the same id? For example, many SetNeedsStyleRecalc calls may flow into a single updateStyle slice.

@natduca Should we try to use frameblamr here? Should we schedule a meeting to sort out how to use flow events here?

**Comment by natduca** [comment]
Mmm good insight. I think you're on the right track, but if you dig deep you'll find that frameblamr isn't the full story. Namely, frameblamr will let us connect a given *slice* to a given execution context. But, there is no way to connect an async event to a slice... if we had that, then it'd work. How would we do that?


E.g. inside the recalcStyle, the animation event is going to get created. Frameblamr is going to tell us the context for the recalcStyle [hopefully, @skyostil am I batty?] But, the animation engine running is what is going to create the async slice... we'll need to figure out how create a connection in the trace (or some new implicit rule) that says that the async event has the contextid of that frame...

**Comment by skyostil** [comment]
> E.g. inside the recalcStyle, the animation event is going to get created. Frameblamr is going to tell us 
> the context for the recalcStyle [hopefully, @skyostil am I batty?]

Yeah, we should be able to know which frame's style is being recalc'd.

> But, the animation engine running is what is going to create the async slice... we'll need to figure out 
> how create a connection in the trace (or some new implicit rule) that says that the async event
> has the contextid of the enclosing slice...

The naive idea I had was to make async events inherit the context of the thread that started them (at the time they were started). That wouldn't draw an arrow between them in the trace but at least the context would match.

**Comment by natduca** [comment]
@dstockwell fyi... do you think when we're recalcing style, we could have a separate trace_event for the work for each frame? or is recalc a big flat thing that doesn't split easily into frames?

**Comment by dstockwell** [comment]
@nduca, I think this is already split. Do you have an example page or trace?

The style recalc event (Document::updateStyle) should strictly belong to the frame referenced by the outer UpdateLayoutTree event.

**Comment by natduca** [comment]
I'm just making assumptions here! What happens if you've done something that invalidated the universe from mainframe down, but then ask for computed style on something in a subframe? Do you see recalcs for each frame visited?

**Comment by benshayden** [comment]
I have no concrete plans to do this until a benchmark requires it, and then I'll probably ask the benchmark author to implement it, so unassigning, but I'll keep this open for reference.
